### ISSUE TITLE
Terragrunt cache for providers doesn't work with nested modules

### ISSUE THREAD

**Issue by mglotov** [gruntwork-io/terragrunt#3050]
**Describe the bug**
We started testing https://terragrunt.gruntwork.io/docs/features/provider-cache/ and faced a problem. 
Terragrunt with enable caching fails if we use "nested" configuration for modules, when one module sources another one. 
For example, to create an AWS VPC we use our own local module that does some extra work and sources a module `terraform-aws-modules/vpc/aws`. In this case we get the next error:
```
INFO[0000] Terragrunt Cache server is listening on 127.0.0.1:54708
INFO[0000] Start Terragrunt Cache server
╷
│ Error: Error accessing remote module registry
│
│   on main.tf line 10:
│   10: module "vpc" {
│
│ Failed to retrieve available versions for module "vpc" (main.tf:10) from
│ registry.terraform.io: host registry.terraform.io does not provide a
│ modules service.
╵

╷
│ Error: Error accessing remote module registry
│
│   on main.tf line 85:
│   85: module "vpc_gateway_endpoints" {
│
│ Failed to retrieve available versions for module "vpc_gateway_endpoints"
│ (main.tf:85) from registry.terraform.io: host registry.terraform.io does
│ not provide a modules service.
╵
```

If we directly call the `terraform-aws-modules/vpc/aws` from the `terragrunt.hcl` file, everything works well
```
$ TERRAGRUNT_PROVIDER_CACHE=1 TERRAGRUNT_PROVIDER_CACHE_DIR=/Users/nikro/.terraform.d/plugin-cache terragrunt plan                                                         <aws:md-sandbox-admin>
INFO[0000] Terragrunt Cache server is listening on 127.0.0.1:57086
INFO[0000] Start Terragrunt Cache server
INFO[0000] Downloading Terraform configurations from tfr:///terraform-aws-modules/vpc/aws?version=5.7.1 into /Users/example/test-repo/....../.terragrunt-cache/QrEmt8mAQ5UY76kC5XOncjM8wYA/ThyYwttwki6d6AS3aD5OwoyqIWA
INFO[0007] Provider "registry.terraform.io/hashicorp/aws/5.45.0" is cached

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Finding hashicorp/aws versions matching ">= 5.30.0"...
- Installing hashicorp/aws v5.45.0...
- Installed hashicorp/aws v5.45.0 (unauthenticated)

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.
```

**To Reproduce**
Steps to reproduce the behavior, code snippets and examples which can be used to reproduce the issue.

```hcl
include "root" {
  path   = find_in_parent_folders()
  expose = true
}

include "env" {
  path   = find_in_parent_folders("env.hcl")
  expose = true
}

generate "providers_versions" {
  path      = "versions.tf"
  if_exists = "overwrite"
  contents  = <<EOF
terraform {
  required_version = ">= 1.7.0"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "${include.root.locals.tf_providers.aws}"
    }
  }
}
EOF
}

terraform {
  source = "${get_path_to_repo_root()}/terraform//modules/aws-vpc"
}

inputs = {
  name               = include.env.locals.name
  cidr               = include.env.locals.values.vpc_cidr
  azs                = include.env.locals.values.azs
  single_nat_gateway = include.env.locals.values.single_nat_gateway
}

```

terraform//modules/aws-vpc/main.tf
```hcl
...
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "5.7.1"

  name = var.name
  cidr = var.cidr

....           
```

**Expected behavior**
Working terragrunt and cache

**Nice to have**
- [ ] Terminal output
- [ ] Screenshots

**Versions**
- Terragrunt version: 0.56.5
- Terraform version: 1.7.3
- Environment details (Ubuntu 20.04, Windows 10, etc.): MacOS Sonoma

**Additional context**

**Comment by pysiekytel** [comment]
I'm experiencing the same issue, I was not able to workaround it

**Comment by jleight** [comment]
I was able to [hack in a fix](https://github.com/jleight/terragrunt/commit/d7285f8acf28449fd708a163fdfba4f864bb8fa7) and use a locally-built terragrunt to get around this issue. I doubt this is a proper fix, but maybe it'll help towards figuring out an actual solution.

**Comment by levkohimins** [comment]
Resolved in [v0.57.3](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.57.3) release.
